### PR TITLE
Definition Graph Visualization for noWorkflow

### PR DESCRIPTION
**Pull request by JoshuaGlaZ**
# Google Summer Code 2024 Final Report

This pull request marks the final submission for my Google Summer of Code (GSoC) 2024 project, "Verify the reproducibility of an experiment". The primary objective of this project was to identify discrepancies in a trial's provenance data and compare them. The current implementation can differentiate trials using the `now vis` application, but it only does so through the trial's activations.

`now vis` command now includes a new feature for displaying a trial's definition graph. Here, the definition refers to the trial's code_component representation and its relationships based on the trial's composition. This graph provides a visual representation of the trial's definition, structured as a tree similar to an Abstract Syntax Tree (AST). 

`now ast` command has been introduced to represent a generated AST structure of a trial. The `now ast` result is using ast module, but it is based on the trial's code_components and compositions rather than the original script. The tool supports two output formats: JSON and Graphviz dot format.

`now diff` command now includes an option to display discrepancies between trials, either for activations or definitions. The differences are displayed using the APTED algorithm. To display the output of trial differences in Tree Edit Distance, simply add `-d` for definition and `-a` for activations.

### CODE REVIEW

File: capture/noworkflow/now/models/definition.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    REPLACE = {
+        "graph_width": "graph.width",
+        "graph_height": "graph.height",
+        "graph_mode": "graph.mode"
+    }
+
+    def __init__(self, trial_ref, **kwargs):
+        super(Definition, self).__init__(trial_ref, **kwargs)
+        self.trial = Trial(trial_ref)
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/definition.py:49]: In some situations we may have the Trial object instead of the trial_ref. It would be good to be able to initialize the Definition both by passing a trial_ref or a Trial object and have this object without reinstantiating Trial

File: capture/noworkflow/now/models/definition_diff.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self, trial_ref1, trial_ref2, **kwargs):
+        super(DefinitionDiff, self).__init__(trial_ref1, trial_ref2, **kwargs)
+        self.definition1 = Definition(trial_ref1)
+        self.definition2 = Definition(trial_ref2)
+
+        self.graph = DefinitionDiffGraph(self)
+        self.initialize_default(kwargs)
+
+    @property
+    def trial(self):
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/definition_diff.py:59]: Is this function based only on the definition? It looks a lot like the funcion for the general diff.

File: capture/noworkflow/now/models/definition_diff.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        """Return a tuple with information from both trials """
+        extra = ("start", "finish", "duration_text", "code_hash")
+        ignore = ("id",)
+        return diff_dict(
+            self.definition1.trial.to_dict(
+                ignore=ignore, extra=extra),                     # pylint: disable=no-member
+            self.definition2.trial.to_dict(ignore=ignore, extra=extra))                     # pylint: disable=no-member
+
+    @property
+    def modules(self):
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/definition_diff.py:69]: Same comment as before. It does not seem to diff only the modules that appear in the definition (and it is not necessary to do so for this pull request - it would require a lot of different work)

File: capture/noworkflow/now/models/definition_diff.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            set(self.definition1.trial.environment_attrs),
+            set(self.definition2.trial.environment_attrs))
+
+    @property
+    def file_accesses(self):
+        """Definition diff file accesses"""
+        return diff_set(
+            set(self.definition1.trial.file_accesses),
+            set(self.definition2.trial.file_accesses),
+            create_replaced=False)
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/definition_diff.py:88]: Same thing. Remove all the functions that are not related to DefinitionDiff.

Or, instead of defining a new class, maybe this definition diff class could be a method in the [Diff](https://github.com/gems-uff/noworkflow/blob/e5cf78953954a471dd6ea3ad4468017b6d3a9363/capture/noworkflow/now/models/diff.py#L18) class.

File: capture/noworkflow/now/models/graphs/definition_diff_graph.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+from apted import meta_chained_config, Config, APTED
+from future.utils import viewitems
+
+from .trial_graph import Node
+from .structures import Graph, prepare_cache
+from .diff_graph import CONFIG, children_cmp, create_mapping, merge, merge_edges
+
+
+def create_diff(trial_graph1, trial_graph2):
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/graphs/definition_diff_graph.py:23]: This function looks the same as the one in [diff_graph.py](https://github.com/gems-uff/noworkflow/blob/e5cf78953954a471dd6ea3ad4468017b6d3a9363/capture/noworkflow/now/models/graphs/diff_graph.py#L151). Is there a difference?

If there is a difference, try to specialize the existing function with optional parameters. Otherwise, just use the existing one.

Avoid copying and pasting the code because it makes it harder do maintain both versions in the future.

File: capture/noworkflow/now/models/graphs/definition_diff_graph.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    finished = finished1 and finished2
+    return finished, graph, nodes
+
+
+cache = prepare_cache(  # pylint: disable=invalid-name
+    lambda self, *args, **kwargs: "definition diff {}:{}".format(self.diff.definition1.trial.id,
+                                                      self.diff.definition2.trial.id))
+
+
+class DefinitionDiffGraph(Graph):
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/graphs/definition_diff_graph.py:96]: Minor issue: in the sense of reducing code repetitions and redefinitions, this could be a mode in the existing [DiffGraph](https://github.com/gems-uff/noworkflow/blob/e5cf78953954a471dd6ea3ad4468017b6d3a9363/capture/noworkflow/now/models/graphs/diff_graph.py#L223)

File: capture/noworkflow/now/persistence/models/code_component.py [hunk 7]
```diff
@@ -136,3 +138,35 @@ def first_evaluation(self):
         """
         from .evaluation import Evaluation
         return self.evaluations.order_by(Evaluation.m.checkpoint).first()
+
+    @classmethod # query
+    def compositions(cls, trial_id):
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/persistence/models/code_component.py:143]: The name of this method hints that it returns all compositions, but it is restricted to the ones used in the definition graph that represents the ast. We should think about a new name. Maybe ast_composition?

File: capture/noworkflow/now/models/definition.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    @property
+    def environment(self):
+        """Diff environment variables"""
+        return self.trial.environment_attrs
+
+    @property
+    def file_accesses(self):
+        """Diff file accesses"""
+        return self.trial.file_accesses
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/models/definition.py:67]: Why does the Definition have these methods?

File: capture/noworkflow/now/vis/views.py [hunk 9]
```diff
@@ -336,7 +338,11 @@ def get_file(file_hash, file_ext):
 @app.route("/trials/<tid>/<graph_mode>/<cache>.json")
 def trial_graph(tid, graph_mode, cache,expCode=None):
     """Respond trial graph as JSON"""
-    trial = Trial(tid)
+    if "definition" in graph_mode :
+        trial = Definition(tid)
+        graph_mode = graph_mode.split('_')[1]
+    else:
+        trial = Trial(tid)
```
Comment (JoaoFelipe) [inline capture/noworkflow/now/vis/views.py:345]: If we had Definition as a graph_mode of Trial as suggested before, this wouldn't be required.

**Review by JoaoFelipe — CHANGES_REQUESTED** [review]
Hi, I left some comments on the files. Can you check if it is feasible to do these changes?

File: capture/noworkflow/now/models/definition.py (hunk 1 above)
Comment (JoshuaGlaZ) [reply capture/noworkflow/now/models/definition.py:49]: Is it necessary that trial_ref (or kwargs) provides a Trial object? Because when calling the trial's ID, I use `self.trial.trial.id`.

**Comment by JoshuaGlaZ** [comment]
I've made the definiton_tree to be a new mode for trial_graph. Don't know if `definition.py` is needed since I combine it with `trial_graph.py`.

As for trial_ast (represents a real AST from code_components + compositions), I moved it to models/ast

File: capture/noworkflow/now/models/definition.py (hunk 1 above)
Comment (JoshuaGlaZ) [reply capture/noworkflow/now/models/definition.py]: As for now, I moved the definition into trial_graph

**Comment by JoaoFelipe** [comment]
Everything looks good in the source code! I'll try to run it here to check for other issues.

Thanks!

**Comment by JoaoFelipe** [comment]
Oh, I have two additional requests. First, can you add a section in the README.md that briefly describes the definition graph and possibility of diff, with figures.


Second: before creating the now vis definition graph, you had a graphviz version, right? Is it working? Maybe you can submit another pull request for a command line that exports it. Like, `now export --definition`

**Comment by JoshuaGlaZ** [comment]
Okay, I've made local changes for the now ast command. The README also mentions the definition graph. Does the README look good?

**Comment by JoaoFelipe** [comment]
Merged! It also fixes #164 and #163

Thanks!
